### ISSUE TITLE
lib curl compiled on OSX 10.13 failed to run on OSX 10.10

### ISSUE THREAD

**Issue by arun-govindan** [curl/curl#2905]
I did this:
I compiled my application with curl (options : ./configure --with-gnutls) on Xcode 9.4. And when I tested the application on Mac OS version yosemite (OS X 10.10) the app crashed with the below output.

Crash report : 
Dyld Error Message:
  Symbol not found: _clock_gettime
  Referenced from: */libcurl.4.dylib (which was built for Mac OS X 10.13)
  Expected in: /usr/lib/libSystem.B.dylib

I expected the following:
Application to run normally

curl/libcurl version:
curl-7.40.0

Mac OS version: yosemite (OS X 10.10)

**Comment by bagder** [comment]
`clock_gettime` apparently wasn't added until 10.12. I don't know how this should be fixed properly (googling "clock_gettime macOS" shows this and related issues have bitten many projects but I found no clear solution).

## Work-around

After you run configure, but before you invoke 'make', you can edit `lib/curl_config.h` and make sure `HAVE_CLOCK_GETTIME_MONOTONIC` is not defined. Then build.

**Comment by nickzman** [comment]
I think I know how to fix this. I looked at the time.h header, and noticed that the clock_* functions are all marked for weak-linking as of macOS 10.13.x, so there are two ways we can weak-link the symbol:

1. Use [traditional weak-linking](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/MachOTopics/1-Articles/executing_files.html#//apple_ref/doc/uid/TP40001829-98432-TPXREF120), which means checking to see if the symbol is not equal to null.
2. Use `__builtin_available()` to check if the user is using macOS 10.12 or later, or iOS 10.0 or later. We have the `HAVE_BUILTIN_AVAILABLE` preprocessor macro to check on this.

Method 2 is the method Apple wants developers to use from now on, but as we found out a while ago, it doesn't work with compilers other than Clang.
